open-craft/xblock-wistia
# feat: implements retry when retrieving captions

## Change description

The initial implementation was causing an infinite loop because the `captionsError` wasn't really incrementing. The variable was being accessed at two different scopes, the function's and the `onReady` function call.

This implementation fixes the infinite loop that was causing webpages to crash. In addition, it refines the implementation of the retry method. 

I was initially hesitant about using try/catch inside the loop statement. 

The performance doesn't take much of a hit. It definitely does, because we're assuming that we're throwing 3 errors in the worst case scenario. However, it's justifiable for what I'm about to explain next.

To identify that the captions plugin was successful, we need a variable to keep track if there was an error or not. We can use said variable to store success values as well, doesn't matter how exactly. 

Storing the variable can be done in two ways, inside the function's context, which was the case initially. This wouldn't work for the reason explained earlier, different scopes.

This brings us to our second method of storing the variable, under the window object.  The window would maintain the variable regardless the scope. We can simply implement something like `window._wq_state` and store the `captionsError` inside it. 

However, we expect that we will have more than one Wistia XBlock in a single page. Accordingly, we'll need to add define each state with a key. Luckily, the initializer already contains an awesome key called `mediaId`. However, that is defined in the local scope of the function similarly to how `captionsError` was initially defined. Accordingly, this will also not work as expected. 

In order to limit the changes needed to the javascript, I chose to move forward with the try/catch block inside the loop.

## Type of change

> This fixes a bug in a new feature, so both 🤷🏼 

## Related issues

* Jira Ticket: SE-4826

## Testing Instructions

* SSH into `test-esme-oxford-koa2.opencraft.hosting`
To access the app server, you'll want to check Ocim

* Run the following code to install the old XBlock version
Remember to wait a little so the services restart
```bash
sudo -Hu edxapp bash && /edx/bin/supervisorctl restart all

cd ~ && . edxapp_env && . venvs/edxapp/bin/activate && pip uninstall wistiavideo-xblock && pip install git+https://github.com/open-craft/xblock-wistia.git@gabor/apply-captions-workaround#egg=wistiavideo_xblock
```

* Open [this subsection of the course](https://studio.test-esme-oxford-koa2.opencraft.hosting/container/block-v1:OpenCraft+OC_001+OCR_001+type@vertical+block@f3fa53d117ab4d619a7f49d1d293bb50) **in chrome**.
Firefox will keep consuming memory, chrome wouldn't.

* SSH again to install the new XBlock version
Again, remember to wait a little so the services restart
```bash
sudo -Hu edxapp bash && /edx/bin/supervisorctl restart all

cd ~ && . edxapp_env && . venvs/edxapp/bin/activate && pip uninstall wistiavideo-xblock && pip install git+https://github.com/open-craft/xblock-wistia.git@nizar/fix-captions-workaround-bug#egg=wistiavideo_xblock
```

* Open [the same subsection of the same course](https://studio.test-esme-oxford-koa2.opencraft.hosting/container/block-v1:OpenCraft+OC_001+OCR_001+type@vertical+block@f3fa53d117ab4d619a7f49d1d293bb50).
You should be able to edit the Wistia XBlock now.

## Checklists

### Development


### Security


### Code review
